### PR TITLE
Keep AI from shooting while facing the wrong direction

### LINKED ISSUE #3938: AI shoots toxin too early (while still turning)
**Important**
Please only submit technical bug reports here. This is not a place to make feature requests
or suggestions. For that please visit our suggestions site:
https://suggestions.revolutionarygamesstudio.com/

Or our community forums for longer discussions
https://community.revolutionarygamesstudio.com/

**Describe the bug**

Currently when the AI decides to fire toxin at something, it chooses to look at the target then immediately queues a shot (https://github.com/Revolutionary-Games/Thrive/blob/master/src/microbe_stage/MicrobeAI.cs#L734) After the reduction in turning speed, this often results in AI species firing toxin wildly into the void. I suspect that this can be fixed organically by comparing the normalized lookAtPoint of the microbe with the facing direction.

**To Reproduce**
Steps to reproduce the behaviour:
1. Swim behind something that can and will shoot toxin at you

**Expected behaviour**
AI microbe turns, then shoots toxin when facing the target.

**Actual behavior**
AI fires as it starts to turn, then finishes turning towards the target.

### PR DESCRIPTION
**Pull request by adQuid**
**Brief Description of What This PR Does**

Modifies AI to not shoot toxin until facing close to the target. The AI still misses frequently, but with the randomness of target turning it's accurate enough to be threatening again.

**Related Issues**

Closes #3938 

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: src/microbe_stage/MicrobeAI.cs [hunk 1]
```diff
@@ -740,7 +740,10 @@ private void LaunchToxin(Vector3 target)
             if (CanShootToxin())
             {
                 microbe.LookAtPoint = target;
-                microbe.QueueEmitToxin(oxytoxy);
+
+                // hold fire until the target is lined up.
+                if (microbe.FacingDirection().Normalized().AngleTo(microbe.LookAtPoint.Normalized()) < 0.1f + SpeciesActivity / (5 * SpeciesFocus))
+                    microbe.QueueEmitToxin(oxytoxy);
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeAI.cs:746]:
```suggestion
                // Hold fire until the target is lined up.
                if (microbe.FacingDirection().Normalized().AngleTo(microbe.LookAtPoint.Normalized()) < 
                    Constants.MICROBE_AI_TOXIN_FIRE_ANGLE_THRESHOLD + SpeciesActivity / (5 * SpeciesFocus))
                {
                    microbe.QueueEmitToxin(oxytoxy);
                }
```

A bit better formatting that should also please the linter (hopefully, I wrote this on Github), and also I suggest adding a tweak constant.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
I don't see from the code why this wouldn't work. I have one suggestion to get the formatting right (hopefully) and add a tweak constant to make it slightly possible to tweak the angle by a designer.

File: src/microbe_stage/MicrobeAI.cs [hunk 2]
```diff
@@ -740,7 +740,11 @@ private void LaunchToxin(Vector3 target)
             if (CanShootToxin())
             {
                 microbe.LookAtPoint = target;
-                microbe.QueueEmitToxin(oxytoxy);
+
+                // hold fire until the target is lined up.
+                if (microbe.FacingDirection().Normalized().AngleTo(microbe.LookAtPoint.Normalized()) <
+                    0.1f + SpeciesActivity / (Constants.AI_BASE_TOXIN_SHOOT_ANGLE_PRECISION * SpeciesFocus))
+                    microbe.QueueEmitToxin(oxytoxy);
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeAI.cs:747]:
```suggestion
                // Hold fire until the target is lined up.
                if (microbe.FacingDirection().Normalized().AngleTo(microbe.LookAtPoint.Normalized()) <
                    0.1f + SpeciesActivity / (Constants.AI_BASE_TOXIN_SHOOT_ANGLE_PRECISION * SpeciesFocus))
                 {
                    microbe.QueueEmitToxin(oxytoxy);
                 }
```
I add braces when there are multiline if conditions to make it visually easier to see what's inside the if statement and what is part of the condition.

**Comment by adQuid** [comment]
this still shows updates in the submodules, but it looks like it's just a commit from earlier today that hasn't made it into master yet.

**Comment by hhyyrylainen** [comment]
Yeah, that's exactly what it is, so it is fine to merge that update through this PR rather than me doing it somewhere else.

File: src/microbe_stage/MicrobeAI.cs (hunk 2 above)
Comment (adQuid) [reply src/microbe_stage/MicrobeAI.cs:747]: That's also my preference. I must have missed that.

Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeAI.cs]: Let's hope I didn't make a mistake with the GitHub editor... If this passes checks then this is fine for merging as long as no one else wants to playtest this.

**Comment by hhyyrylainen** [comment]
Nope, the indent sizes ended up incorrect. Do you want to correct it or should I?

**Comment by adQuid** [comment]
> Nope, the indent sizes ended up incorrect. Do you want to correct it or should I?

done

**Review by hhyyrylainen — APPROVED** [review]
I'll approve this now but I'll wait a day or two before merging in case someone wants to playtest this.

**Pull request merged**